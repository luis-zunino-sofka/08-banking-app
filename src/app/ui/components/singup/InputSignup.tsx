/* eslint-disable @typescript-eslint/ban-ts-comment */
import { SignupFormData } from "@core/interfaces";
import { FieldErrors, UseFormRegisterReturn } from "react-hook-form";

interface IInput {
  register: UseFormRegisterReturn<string>;
  name: string;
  type: string;
  label: string;
  errors: FieldErrors<SignupFormData>;
  placeHolder: string;
}

export const InputSignup = (props: IInput) => {
  const { register, name, type, label, errors, placeHolder } = props;
  return (
    <div className="signup-form__input-group">
      <label htmlFor={`input-${name}`} className="signup-form__label">
        {label}
      </label>
      <input
        id={`input-${name}`}
        type={type}
        className="signup-form__input"
        placeholder={placeHolder}
        {...register}
      />
      {/* @ts-expect-error */}
      {errors[name] && (
        /* @ts-expect-error */
        <span className="error-message">{errors[name].message}</span>
      )}
    </div>
  );
};
