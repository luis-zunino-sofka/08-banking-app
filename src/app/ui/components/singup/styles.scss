@use "../../styles/variables" as var;

.signup-form {
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #f4f7fa;
  font-family: Arial, sans-serif;

  &__form {
    background-color: var.$bg-color-white;
    height: Calc(100vh - 20px);
    padding: 30px;
    border-radius: 8px;
    box-shadow: 0 4px 20px rgba(0, 0, 0, 0.1);
    width: 100%;
    max-width: fit-content;
    box-sizing: border-box;
  }

  &__title {
    font-size: 28px;
    font-weight: bold;
    margin-bottom: 20px;
    text-align: center;
  }

  &__input-group {
    margin-bottom: 20px;
    display: flex;
    flex-direction: column;
  }

  &__label {
    font-size: 16px;
    font-weight: 600;
    margin-bottom: 8px;
    display: block;
  }

  &__input {
    width: 250px;
    padding: 12px;
    font-size: 16px;
    border: 1px solid #ccc;
    border-radius: 5px;
    box-sizing: border-box;
    transition: border 0.3s ease;

    &:focus {
      border-color: #007bff;
      outline: none;
    }
  }

  &__error {
    color: var.$text-color-error;
    font-size: 12px;
    margin-top: 5px;
  }

  &__submit {
    background-color: #007bff;
    color: white;
    font-size: 16px;
    font-weight: bold;
    padding: 12px;
    width: 100%;
    border: none;
    border-radius: 5px;
    cursor: pointer;
    transition: background-color 0.3s ease;
    align-self: center;
    margin: auto;
    &:hover {
      background-color: #0056b3;
    }

    &:disabled {
      background-color: #ccc;
      cursor: not-allowed;
    }
  }

  &__footer {
    text-align: center;
    margin-top: 20px;
    display: flex;
    gap: 1rem;
    justify-content: center;
  }

  &__login-link {
    color: #007bff;
    text-decoration: none;

    &:hover {
      text-decoration: underline;
    }
  }

  .flex {
    display: flex;
    gap: 10px;
  }
}
