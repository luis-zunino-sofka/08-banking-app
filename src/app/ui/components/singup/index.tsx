import { useSignup } from "@core/hooks/signup/useSignup";
import { ROUTE_PATH } from "@core/constants";
import { Toaster } from "react-hot-toast";
import "./styles.scss";
import { InputSignup } from "./InputSignup";

export const SignupForm = () => {
  const { errors, isLoading, onSubmit, register, handleSubmit, watch } =
    useSignup();

  return (
    <div className="signup-form">
      <form
        id="register-form"
        className="signup-form__form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <h2 className="signup-form__title">Regístrate</h2>
        <div className="flex">
          <InputSignup
            register={register("name", {
              required: "El nombre es obligatorio",
            })}
            name="name"
            type="text"
            label="Nombre:"
            errors={errors}
            placeHolder="Ingrese su nombre"
          />

          <InputSignup
            register={register("lastName", {
              required: "El apellido es obligatorio",
            })}
            name="lastName"
            type="text"
            label="Apellido:"
            errors={errors}
            placeHolder="Ingrese su apellido"
          />
        </div>
        <div className="flex">
          <InputSignup
            register={register("dni", { required: "El DNI es obligatorio" })}
            name="dni"
            type="text"
            label="DNI:"
            errors={errors}
            placeHolder="Ingrese su dni"
          />

          <InputSignup
            register={register("username", {
              required: "El nombre de usuario es obligatorio",
            })}
            name="username"
            type="text"
            label="Nombre de usuario:"
            errors={errors}
            placeHolder="Ingrese su nombre de usuario"
          />
        </div>
        <div className="flex">
          <InputSignup
            register={register("password", {
              required: "El nombre de usuario es obligatorio",
            })}
            name="password"
            type="password"
            label="Contraseña:"
            errors={errors}
            placeHolder="Contraseña"
          />

          <InputSignup
            register={register("confirmPassword", {
              required: "La confirmación de la contraseña es obligatoria",
              validate: (value) =>
                value === watch("password") || "Las contraseñas no coinciden",
            })}
            name="confirmPassword"
            type="password"
            label="Confirmar contraseña:"
            errors={errors}
            placeHolder="Confirmar contraseña"
          />
        </div>
        <button
          type="submit"
          className="signup-form__submit"
          disabled={isLoading}
        >
          Registrarse
        </button>
        <div className="signup-form__footer">
          <p>
            ¿Ya tienes cuenta?
            <a
              href={ROUTE_PATH.LOGIN}
              id="login-link"
              className="signup-form__login-link"
            >
              Inicia sesión aquí
            </a>
          </p>
        </div>
      </form>
      <Toaster />
    </div>
  );
};
