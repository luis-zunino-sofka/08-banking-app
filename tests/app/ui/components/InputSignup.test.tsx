import { render, screen, fireEvent } from "@testing-library/react";
import { describe, expect, test, vi } from "vitest";
import { FormProvider, useForm, UseFormRegisterReturn } from "react-hook-form";
import React, { ReactNode } from "react";
import { InputSignup } from "../../../../src/app/ui/components/singup/InputSignup";

export const renderWithReactHookForm = (
  ui: ReactNode,
  { defaultValues = {} } = {}
) => {
  const Wrapper = ({ children }) => {
    const methods = useForm({ defaultValues });
    return <FormProvider {...methods}>{children}</FormProvider>;
  };

  return {
    ...render(ui, { wrapper: Wrapper }),
  };
};

describe("InputSignup", () => {
  const mockRegister: UseFormRegisterReturn<string> = {
    onChange: vi.fn(),
    onBlur: vi.fn(),
    name: "name",
    ref: vi.fn(),
  };

  const mockErrors = {};

  test("Match snapshoot", () => {
    const { asFragment } = render(
      <InputSignup
        register={mockRegister}
        name="name"
        type="name"
        label="Name"
        errors={mockErrors}
        placeHolder="Enter your name"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("debe realizar la entrada con la etiqueta y el marcador de posición correctos", () => {
    renderWithReactHookForm(
      <InputSignup
        register={mockRegister}
        name="name"
        type="name"
        label="Name:"
        errors={mockErrors}
        placeHolder="Enter your name"
      />,
      { defaultValues: {} }
    );
    expect(screen.getByText("Name:")).toBeInTheDocument();

    expect(screen.getByPlaceholderText("Enter your name")).toBeInTheDocument();

    expect(screen.getByPlaceholderText("Enter your name")).toHaveAttribute(
      "type",
      "name"
    );
  });

  test("debe mostrar un mensaje de error cuando se pasan los errores", () => {
    const errors = { name: { message: "Email is required" } };

    renderWithReactHookForm(
      <InputSignup
        register={mockRegister}
        name="name"
        type="name"
        label="Name:"
        errors={errors}
        placeHolder="Enter your name"
      />,
      { defaultValues: {} }
    );

    expect(screen.getByText("Email is required")).toBeInTheDocument();
  });

  test("debe llamar a la función de registro en el cambio de entrada", () => {
    renderWithReactHookForm(
      <InputSignup
        register={mockRegister}
        name="name"
        type="name"
        label="Name:"
        errors={mockErrors}
        placeHolder="Enter your name"
      />,
      { defaultValues: {} }
    );

    const inputElement = screen.getByPlaceholderText("Enter your name");
    fireEvent.change(inputElement, { target: { value: "test@example.com" } });

    expect(mockRegister.onChange).toHaveBeenCalled();
  });

  test("no debe mostrar un mensaje de error cuando no se pasan errores", () => {
    renderWithReactHookForm(
      <InputSignup
        register={mockRegister}
        name="name"
        type="name"
        label="Email"
        errors={mockErrors}
        placeHolder="Enter your name"
      />
    );

    expect(screen.queryByText("Email is required")).toBeNull();
  });

  test("debe guardar el valor cuando se le ingresa un valor en el input", async () => {
    renderWithReactHookForm(
      <InputSignup
        register={mockRegister}
        name="name"
        type="name"
        label="Name:"
        errors={mockErrors}
        placeHolder="Enter your name"
      />,
      { defaultValues: {} }
    );
    const nombre = screen.getByRole("textbox", { name: "Name:" });

    fireEvent.change(nombre, {
      target: { value: "Pepe" },
    });

    expect(nombre).toHaveValue("Pepe");
  });
});
