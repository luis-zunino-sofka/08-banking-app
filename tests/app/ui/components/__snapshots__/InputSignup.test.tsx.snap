// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputSignup > Match snapshoot 1`] = `
<DocumentFragment>
  <div
    class="signup-form__input-group"
  >
    <label
      class="signup-form__label"
      for="input-name"
    >
      Name
    </label>
    <input
      class="signup-form__input"
      id="input-name"
      name="name"
      placeholder="Enter your name"
      type="name"
    />
  </div>
</DocumentFragment>
`;
